### DISCUSSION TITLE
EstlCam Support

### DISCUSSION THREAD

**Discussion by tlasrich** [TimPaterson/Fusion360-Batch-Post#64] [Q&A] [answered]
Hello Everyone,

so far I have always used EstlCam as CAM for my hobby projects. Due to a problem with a new project, I used Fusion360 as CAM. I quickly became aware of the limitations of the free version. During a search on the internet I ended up here. However, when I use EstlCam as a post processor, I get the following error message:

Failed on setup Setup1: Tool change G-code (Txx) not found; this post processor is not compatible with Post Process All.

Are there any efforts to add EstlCam as a supported post processor? Can I provide support here?

Thx.

**Comment by TimPaterson** [comment]
The EstlCam post doesn't seem to target a specific machine. What machine are you using, and isn't there a Fusion 360 post for it?  Anyway, to use EstlCam with PostProcessAll you would have to modify EstlCam.cps to output the tool change G-code, just like the error message says. Line 338 is where it writes the _comment_ for the tool change, so you would need to have it output it there, maybe
`writeBlock("T" + toolFormat.format(tool.number));`

**Reply by tlasrich** [reply]
Hi,

there is a post processor for EstlCam for Fusion360. With this I can also create the following individual files with the free version:

_%
(1002)
(T36  D=6. CR=0. - ZMIN=-6. - SCHAFTFRSER)
G90 G94 G17 G91.1
G21
(WHEN USING FUSION 360 FOR PERSONAL USE, THE FEEDRATE OF)
(RAPID MOVES IS REDUCED TO MATCH THE FEEDRATE OF CUTTING)
(MOVES, WHICH CAN INCREASE MACHINING TIME. UNRESTRICTED RAPID)
(MOVES ARE AVAILABLE WITH A FUSION 360 SUBSCRIPTION.)
G53 G0 Z0.
(2D-KONTUR1)
M9
T36 M6
S24000 M3
G54
M9
G0 X53.2 Y-0.6
G0 G43 Z15. H36
G1 Z5. F1500.
G1 Z1. F333.33
G1 Z-5.4
G1 X53.207 Z-5.494 F600.
G1 X53.229 Z-5.585
G1 X53.265 Z-5.672
G1 X53.315 Z-5.753
G1 X53.376 Z-5.824
G1 X53.447 Z-5.885
G1 X53.528 Z-5.935
G1 X53.615 Z-5.971
G1 X53.706 Z-5.993
G1 X53.8 Z-6.
G1 X54.4
G3 X55. Y0. I0. J0.6
G3 X-55. I-55. J0. F1500.
G3 X55. I55. J0.
G3 X54.4 Y0.6 I-0.6 J0. F600.
G1 X53.8
G1 X53.706 Z-5.993
G1 X53.615 Z-5.971
G1 X53.528 Z-5.935
G1 X53.447 Z-5.885
G1 X53.376 Z-5.824
G1 X53.315 Z-5.753
G1 X53.265 Z-5.672
G1 X53.229 Z-5.585
G1 X53.207 Z-5.494
G1 X53.2 Z-5.4
G1 Z15. F1500.
M9
G53 G0 Z0.
M30
%_

These files contain corresponding entries for the tool in the header. I use the same postprocessor for the plugin.
I added the post processorfile maybe it helps for a better understamding. Because i'm not shure how to 

Link to EstlCam Command List: https://www.estlcam.de/steuerung_cnc_programme.php
[estlcam 2.txt](https://github.com/TimPaterson/Fusion360-Batch-Post/files/13856767/estlcam.2.txt)

**Comment by TimPaterson** [comment]
I was using the EstlCam.cps provided by Fusion (identical to the one you attached) when I described the required code change.  In my testing, PostProcessAll worked correctly when the line of code I listed above was added after line #338 (which reads `writeComment("Insert tool #" + toolFormat.format(tool.number));`).

The G-code you posted has the "T36" block which is what PostProcessAll needs.  But reading the code in EstlCam.cps, I don't see how it could have produced that listing -- I see no code to output a Txx block unless you add that line of code I suggested.

**Reply by tlasrich** [reply]
Hello Tim,

thanks for your support. I tried to implement your hint, but even after inserting the code after line 338, I still get the same error message.

Is it possible for you to send me the Post Processor file that worked in your test? Thank you.

**Comment by TimPaterson** [comment, marked as answer]
Here the post. Change the extension back to `.cps` to use it.
[estlcam.txt](https://github.com/TimPaterson/Fusion360-Batch-Post/files/13890938/estlcam.txt)

This is a snippet of it's output for a tool change:
```
(Thread1)
(Insert tool #21)
M9 G30
T21
M00
M03 S10000
M08
G00 X0.2683 Y-0.6217
G00 Z0.5
```

The M9 G30 above the T21 is from PostProcessAll's `G-code for tool change` entry.

**Reply by tlasrich** [reply]
Hello Timo,

Thank you very much. I don't know why, but now that I have completely removed the postprocessor and added the file from you, it finally works. At least I don't get an error message. :-D

Thank you, greetings Thorsten
